envi/archs/arm/emu.py

```diff
@@ -164,70 +197,94 @@ def readMemValue(self, addr, size):
-        if size == 1:
-            return struct.unpack("B", bytes)[0]
-        elif size == 2:
-            return struct.unpack("<H", bytes)[0]
-        elif size == 4:
-            return struct.unpack("<I", bytes)[0]
-        elif size == 8:
-            return struct.unpack("<Q", bytes)[0]
+
+        endian_fmt = (LSB_FMT, MSB_FMT)[self.getEndian()]
+        return struct.unpack(endian_fmt[size], bytes)[0]
 
     def writeMemValue(self, addr, value, size):
-        #FIXME change this (and all uses of it) to passing in format...
-        #FIXME: Remove byte check and possibly half-word check.  (possibly all but word?)
-        if size == 1:
-            bytes = struct.pack("B",value & 0xff)
-        elif size == 2:
-            bytes = struct.pack("<H",value & 0xffff)
-        elif size == 4:
-            bytes = struct.pack("<I", value & 0xffffffff)
-        elif size == 8:
-            bytes = struct.pack("<Q", value & 0xffffffffffffffff)
+        endian_fmt = (LSB_FMT, MSB_FMT)[self.getEndian()]
+        mask = e_bits.u_maxes[size]
+        bytes = struct.pack(endian_fmt[size], (value & mask))
         self.writeMemory(addr, bytes)
 
     def readMemSignedValue(self, addr, size):
-        #FIXME: Remove byte check and possibly half-word check.  (possibly all but word?)
         bytes = self.readMemory(addr, size)
         if bytes == None:
             return None
-        if size == 1:
-            return struct.unpack("b", bytes)[0]
-        elif size == 2:
-            return struct.unpack("<h", bytes)[0]
-        elif size == 4:
-            return struct.unpack("<l", bytes)[0]
+        if len(bytes) != size:
+            raise Exception("Read Gave Wrong Length At 0x%.8x (va: 0x%.8x wanted %d got %d)" % (self.getProgramCounter(),addr, size, len(bytes)))
+
+        endian_fmt = (LSB_FMT_SIGNED, MSB_FMT_SIGNED)[self.getEndian()]
+        return struct.unpack(endian_fmt[size], bytes)[0]
+
+    def parseOpcode(self, va, arch=envi.ARCH_DEFAULT):
+        '''
+        Parse an opcode from the specified virtual address.
+
+        Example: op = m.parseOpcode(0x7c773803)
+
+        note: differs from the IMemory interface by checking loclist
+        '''
+        if arch == envi.ARCH_DEFAULT:
+            tmode = self.getFlag(PSR_T_bit)
+            arch = (envi.ARCH_ARMV7, envi.ARCH_THUMB)[tmode]
+
+        off, b = self.getByteDef(va)
+        return self.imem_archs[ (arch & envi.ARCH_MASK) >> 16 ].archParseOpcode(b, off, va)
 
     def executeOpcode(self, op):
         # NOTE: If an opcode method returns
         #       other than None, that is the new eip
-        x = None
-        if op.prefixes >= 0xe or conditionals[op.prefixes](self.getRegister(REG_FLAGS)>>28):
-            meth = self.op_methods.get(op.mnem, None)
-            if meth == None:
-                raise envi.UnsupportedInstruction(self, op)
-            x = meth(op)
+        try:
+            self.setMeta('forrealz', True)
```

Comment:
certain ARM operand types actually update the index register on execution.  this is a way to have that functionality only happen when emulating (instead of other one-off "emu.executeOp(op)" calls which can be used for various analysis modules.  
an alternative would be to introduce a new api  that the emulator calls after every instruction from stepi or runFunction.  that would cost performance, however.

i'm ok with changing the meta-name 'forrealz', or commenting it so it's clear.  however, i think this approach is the best one.